### DISCUSSION TITLE
Importing owl models into SADL files

### DISCUSSION THREAD

**Discussion by AbhaMoitra** [SemanticApplicationDesignLanguage/sadl#624] [General]
I was trying to import an owl file https://github.com/idaholab/DIAMOND/blob/master/diamond.owl into a SADL project but ran into issues. I then tried to import a much simpler (and you may say a well-known) owl file https://protege.stanford.edu/ontologies/pizza/pizza.owl but still ran into issues. Any suggestions on what should be done?

I also tried loading a simple owl file from one of our projects into Protege 5.5.0 and was not able to do so. The owl file I tried this on is https://github.com/ge-high-assurance/RACK/blob/master/RACK-Ontology/OwlModels/PROV-S.owl. Now this owl file imports "sadlbasemodel", "sadlimplicitmodel" so they need to be incorporated. Any suggestions on best way to do that?

**Comment by crapo** [comment]
@AbhaMoitra , following the instructions at http://semanticapplicationdesignlanguage.github.io/sadl/ExternalModels.html (which I just updated, it was in the old documentation but not easily found), I have this resulting structure. (I downloaded the file first to my local drive, hence the path shown.)
![image](https://user-images.githubusercontent.com/1186278/110171572-e93e3a00-7dc9-11eb-991e-d8c26df03829.png)
I then created a .sadl file and imported the downloaded external model.
![image](https://user-images.githubusercontent.com/1186278/110171658-0bd05300-7dca-11eb-8910-39a50be07da9.png)
Running the query shown, I got results like:

 `Inference of '/home/camfe/andy/ws/sadl/runtime-EclipseApplication2/GH-624/importdiamond.sadl' requested.<br/>
 Query: select ?c where {?c <rdf:type> <owl:Class>}<br/>
      "c"<br/>
      "e938a9dd-38ad-48cb-b5e0-83d079ed6240(blank node)"<br/>
      "http://sadl.org/sadlimplicitmodel#ExternalEquation"<br/>
      "http://sadl.org/sadlimplicitmodel#GraphPattern"<br/>
      "http://sadl.org/sadlimplicitmodel#GPLiteralValue"<br/>
      "a4954bcb-0406-4683-8493-f430f814e776(blank node)"<br/>
      "http://sadl.org/sadlimplicitmodel#GPVariable"<br/>
      "http://sadl.org/sadlimplicitmodel#DataDescriptor"<br/>
      "http://sadl.org/sadlimplicitmodel#Script"<br/>
      "5b8358ec-b48f-4332-92fd-0979f0bbe78e(blank node)"<br/>
      "http://sadl.org/sadlimplicitmodel#SemanticType"<br/>
      "http://sadl.org/sadlimplicitmodel#AugmentedType"<br/>
      "2d26438e-cb57-4be2-9df0-66fa11c4eb08(blank node)"<br/>
      "http://sadl.org/sadlimplicitmodel#UnittedQuantity"<br/>
      "b41f8bfe-ba05-43f4-b6b3-6bd756ef1dec(blank node)"<br/>
      "http://sadl.org/sadlimplicitmodel#Language"<br/>
      "http://sadl.org/sadlimplicitmodel#GPAtom"<br/>
      "http://sadl.org/sadlimplicitmodel#DataTableRow"<br/>
      "http://sadl.org/sadlimplicitmodel#DataTable"<br/>
      "ff7b6bc6-4ea3-4a87-a612-8a3a9c5d8e6a(blank node)"<br/>
      "http://sadl.org/sadlimplicitmodel#SemanticConstraint"<br/>
      "http://sadl.org/sadlimplicitmodel#GPResource"<br/>
      "http://sadl.org/sadlimplicitmodel#TriplePattern"<br/>
      "http://sadl.org/sadlimplicitmodel#ScientificConcept"<br/>
      "http://sadl.org/sadlimplicitmodel#NamedQuery"<br/>
      "http://sadl.org/sadlimplicitmodel#Equation"<br/>
      "http://sadl.org/sadlimplicitmodel#FunctionPattern"<br/>
      "http://sadl.org/sadlimplicitmodel#Rule"<br/>
      "http://sadl.org/sadllistmodel#List"<br/>
      "f27b2ec8-1870-42d0-8ff7-42aa27093cb3(blank node)"<br/>
      "nuclear:001188"<br/>
      "nuclear:000335"<br/>
      "nuclear:000236"<br/>
      "nuclear:000253"<br/>
      "nuclear:001168"<br/>
      "nuclear:001190"<br/>
      "nuclear:000541"<br/>
      "nuclear:000169"<br/>'

You'll observe an issue here. Namespace display in query results is turned on, as shown by the classes from the SadlImplcitModel. However, the classes from diamond.owl show as "nuclear:dddddd" where d is a digit. The cause is apparent when examining the content of the diamond.owl file. The header declares the xml:base to be "https://github.com/idaholab/DIAMOND/" and concepts in the file are designated, for example, by '<owl:Class rdf:about="https://github.com/idaholab/DIAMOND/nuclear:000703">'. The way I read the W3C standards, this is not a value URI. It certainly confounds the meaning of QNames of the form <prefix>:<localfragment>. The parsers used by Jena don't seem to be able to separate such a URI into namespace and local fragment based on query output, and SADL expects a "#" delimiter between the two. At a minimum, SADL will interpret a reference to "nuclear:000703" to be a QName with prefix "nuclear" and an [invalid since it begins with a digit] local fragment "000703".

**Reply by AbhaMoitra** [reply]
Thanks for the link to "External Model" import; I was not doing it this way. So, will try following the instructions. 

Yes, I had seen "nuclear:01.." etc. and that of course is not legal. Just to proceed I replaced the ":" by "zzz" in the SADL file that I generated.

**Comment by crapo** [comment]
@AbhaMoitra , as far as your second comment about loading into Protege 5.5.0, I'm unsure if Protege requires that imports be available or how it handles that. I would be surprised if the OWL files are not valid as they are written by Jena. You did not identify the issue that you encountered loading into Protege other than saying that you were not able to do so.

**Comment by AbhaMoitra** [comment]
What I saw was the following:
![image](https://user-images.githubusercontent.com/28540315/110174000-70d97800-7dcd-11eb-998e-c21abf5f4062.png)

**Reply by crapo** [reply]
At least it identifies the line and column of the error. In the PROV-S.owl file in github, the line is:
`    xmlns:sadlbasemodel="http://sadl.org/sadlbasemodel#"
`which seems pretty much like the line before and the line after....

Known issue? https://github.com/allysonlister/swo/issues/40

**Comment by AbhaMoitra** [comment]
I followed the instructions and I am now able to query and even graph a small part of it. Thank you.
